### PR TITLE
Import hyperv support from DelphixOS

### PR DESCRIPTION
**Pull request by mvwieringen**
Hi, given Delphix moved from illumos to linux recently I had a quick look of things that might be of interest to import into omniosce. Given the strong boost in virtualization support recently I think it might make sense to import the HYPERV support from DeplhixOS so we have native support on Azure for network and storage analog to the virtio support on other hypervisors. I had a quick look and it seems we only need one commit from DelphixOS which contains the full hyperv stuff. I commited that with the original author and did two small additional patches one that makes things compile on illumos-omnios and one that updates the package manifests to no longer deliver 32 bits drivers as that landed in illumos after Delphix merged the HYPERV support.

The code has its origin in FreeBSD and was ported by Delphix an interesting presentation on the subject was done by George Wilson at BSDCAN 2017

https://www.bsdcan.org/2017/schedule/events/856.en.html

I compiled things and the drivers and daemon compile ok. Packaging looks ok too but have not created packages yet.

### CODE REVIEW

**Comment by citrus-it** [comment]
Hi and thanks for doing this! I know that Delphix have previously built and hosted OmniOS in Azure using their drivers so the chances of this working straight out of the box are good.

Have you contacted @grwilson at all just to make him aware that you're doing this?

**Comment by mvwieringen** [comment]
I was just about to send him an email and have done that now. I first tried if I could get things to compile.

**Comment by citrus-it** [comment]
It looks like there are a handful of other commits needed:

https://github.com/delphix/delphix-os/commits/5.3/usr/src/uts/intel/io/hyperv

Could you add these please?

The last one is where they fixed (removed) the `__packed` attribute so you already have that covered.

**Comment by mvwieringen** [comment]
Ok added those missing commits by cherry-picking them from one repo to an other. Also redid my fixes for __packed by removing both entries just as was done in the big sync commit in DelphixOS. Will compile it later on today and fix any problems if they occur but given the little changes expect no problems.

**Comment by hadfl** [comment]
I built and tested this branch in my hyper-v test environment. It eventually panics with the following trace:

```
> $C
fffffe001030f960 vpanic()
fffffe001030f9f0 cmn_err+0x66(3, fffffffffbede378)
fffffe001030fa30 impl_ddi_free_nodeid+0x9b(8e)
fffffe001030fa60 i_ddi_free_node+0x1d0(fffffe0bd999aaa8)
fffffe001030fab0 ddi_remove_child+0x72(fffffe0bd999aaa8, 0)
fffffe001030fae0 ndi_devi_free+0x2e(fffffe0bd999aaa8)
fffffe001030fb10 vmbus_delete_child+0x6a(fffffe0bd84fe000)
fffffe001030fb40 vmbus_prichan_detach_task+0x18(fffffe0bd84fe000)
fffffe001030fc00 taskq_thread+0x2d0(fffffe0bd995e938)
fffffe001030fc10 thread_start+8()
``` 

```
> ::panicinfo
             cpu                0
          thread fffffe001030fc20
         message impl_ddi_free_nodeid: nodeid 8e already free
```

**Comment by mvwieringen** [comment]
Strange I just checked and usr/src/uts/intel/io/hyperv on both DelphixOS and illumos-omnios is exactly the same. Also checked the coding which calls ndi_devi_free() which is only in two places and both set the handle to NULL afterwards and a check is in place to not call ndi_devi_free() again when the handle is NULL. So currently it beats me what is going on.

**Comment by citrus-it** [comment]
Strange indeed. This is a use-after-free:

```
> fffffe0bd999aaa8::whatis
fffffe0bd999aaa8 is freed from dev_info_node_cache
```

**Comment by hadfl** [comment]
@citrus-it - I did onu to a debug variant and enabled verbose boot. No crash since.

**Comment by citrus-it** [comment]
Let's give it a few more days to see if another crash occurs.

**Comment by mvwieringen** [comment]
Think that is a good idea. I'm also waiting on maybe some input from the origninal author as to if we are doing it the right way e.g. have everything we need to upstream this feature.

**Comment by citrus-it** [comment]
@mvwieringen - let us know when you are happy for this to be merged.

**Comment by mvwieringen** [comment]
Lets wait till Monday 11/03/2019 to give the original author a bit more time to respond with something we might have missed. If nothing is mentioned by Monday I think we can merge and do any followup (if at all needed) later on.

**Comment by mvwieringen** [comment]
@citrus-it - I think we can merge things. I didn't get any feedback but I think we are good anyway.

**Comment by grwilson** [comment]
I went and looked at our commits and you have all but one (a recent one). I will comment in the code.

File: usr/src/uts/intel/io/hyperv/vmbus/vmbus.c [hunk 1]
```diff
@@ -0,0 +1,1453 @@
...
+{
+	int error = 0;
+
+	mutex_enter(&vmbus_lock);
+	if (chan->ch_dev != NULL) {
+		if (ddi_prop_update_string(DDI_DEV_T_NONE, chan->ch_dev,
+		    VMBUS_STATE, VMBUS_STATE_OFFLINE) != DDI_SUCCESS) {
+			dev_err(chan->ch_dev, CE_WARN, "Unable to set "
+			    "\"%s(%s)\" property", VMBUS_STATE,
+			    VMBUS_STATE_OFFLINE);
```
Comment (grwilson) [inline usr/src/uts/intel/io/hyperv/vmbus/vmbus.c]: `mutex_exit(&vmbus_lock)` is missing here

File: usr/src/uts/intel/io/hyperv/vmbus/vmbus.c [hunk 2]
```diff
@@ -0,0 +1,1453 @@
...
+			return (DDI_FAILURE);
+		}
+
+		if (ndi_devi_offline(chan->ch_dev, NDI_DEVI_REMOVE) !=
+		    DDI_SUCCESS) {
+			dev_err(chan->ch_dev, CE_WARN, "Unable to offline "
+			    "device");
```
Comment (grwilson) [inline usr/src/uts/intel/io/hyperv/vmbus/vmbus.c]: mutex_exit(&vmbus_lock) is missing here

**Comment by grwilson** [comment]
It's also worth adding that this was a port from FreeBSD so it might be good for someone to look at new functionality and update this code.

I also have some hacked up `waagent` for OmniOS that I could post if anybody is interested in getting this working on Azure. I had put up an old version of OmniOS into the Azure marketplace.

**Comment by mvwieringen** [comment]
@grwilson - Thanks for your input sorry for not waiting longer. I'm about to fix those last two problems and push a commit with your name as author into this branch. Regarding going back to FreeBSD I might do that when I have some spare time on my hand. Regarding the agent stuff lets try to save as much as possible so if your have something that might be somewhat usable that is better then not having it I would say.

**Comment by citrus-it** [comment]
On Mon, 11 Mar 2019, George Wilson wrote:

; I also have some hacked up `waagent` for OmniOS that I could post if anybody is interested in getting this working on Azure. I had put up an old version of OmniOS into the Azure marketplace.

Hi George,

Are you able to provide a pointer to the waagent? I'm interested in bringing
this into OmniOS and then upstream to illumos.

Thanks,

Andy

**Comment by danmcd** [comment]
He just open-sourced it.  https://github.com/grwilson/WALinuxAgent/tree/omnios

It needs some catching-up with upstream, and maybe some cleaning (some /opt/delphix that might not be strictly needed) BUT it's a good start.  Thanks to @grwilson for making it available.

**Comment by danmcd** [comment]
https://github.com/danmcd/WALinuxAgent/

**Comment by citrus-it** [comment]
https://github.com/omniosorg/omnios-build/pull/1428/files

**Pull request merged**